browser/ui/webui/brave_wallet/image/image_ui.cc

```diff
@@ -0,0 +1,71 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ui/webui/brave_wallet/image/image_ui.h"
+
+#include <string>
+
+#include "brave/components/brave_wallet/browser/brave_wallet_constants.h"
+#include "brave/components/constants/webui_url_constants.h"
+#include "brave/components/l10n/common/locale_util.h"
+#include "brave/components/image_display/resources/grit/image_display_generated_map.h"
+#include "chrome/browser/ui/webui/webui_util.h"
+#include "chrome/grit/browser_resources.h"
+#include "chrome/grit/generated_resources.h"
+#include "components/grit/brave_components_resources.h"
+#include "content/public/browser/web_contents.h"
+#include "content/public/browser/web_ui_data_source.h"
+#include "ui/resources/grit/webui_generated_resources.h"
+
+namespace image {
+
+UntrustedImageUI::UntrustedImageUI(content::WebUI* web_ui)
+    : ui::UntrustedWebUIController(web_ui) {
+  auto* untrusted_source = content::WebUIDataSource::Create(kUntrustedImageURL);
+
+  for (const auto& str : brave_wallet::kLocalizedStrings) {
+    std::u16string l10n_str =
+        brave_l10n::GetLocalizedResourceUTF16String(str.id);
+    untrusted_source->AddString(str.name, l10n_str);
+  }
+
+  untrusted_source->SetDefaultResource(IDR_BRAVE_WALLET_IMAGE_DISPLAY_HTML);
+  untrusted_source->AddResourcePaths(
+      base::make_span(kImageDisplayGenerated, kImageDisplayGeneratedSize));
+  untrusted_source->AddFrameAncestor(GURL(kBraveUIWalletPageURL));
+  untrusted_source->AddFrameAncestor(GURL(kBraveUIWalletPanelURL));
+  webui::SetupWebUIDataSource(
+      untrusted_source,
+      base::make_span(kImageDisplayGenerated, kImageDisplayGeneratedSize),
+      IDR_BRAVE_WALLET_IMAGE_DISPLAY_HTML);
+  untrusted_source->OverrideContentSecurityPolicy(
+      network::mojom::CSPDirectiveName::ScriptSrc,
+      std::string("script-src 'self' chrome-untrusted://resources "
```

Comment:
are you sure both `chrome-untrusted://resources` and `chrome-untrusted://brave-resources` are needed?

Reply:
updated to `script-src 'self' 'unsafe-inline'`

Reply:
Which resources are we looking to rely on here? Could we be more specific about which resources we're wanting to load here. E.g. if we want this resource can we specify it explicitly here? 
`chrome-untrusted://resources/js/load_time_data.m.js`

The concern I have is that a script that executes here is operating as an iframe inside the wallet webui and if we misunderstand the permissions being granted to these iframes a specifically crafted src built into something like an SVG would allow for a script to execute and call into the wallet main frame to trigger a transaction or even worse be able to call over the API layer of the mojo IPC so the goal here is to limit the possible scripts that execute here as much as possible so that we can rely on the CSP as a defense layer in combination with the untrusted frame to prevent this all together.